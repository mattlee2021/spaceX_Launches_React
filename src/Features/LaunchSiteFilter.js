import styles from "./LaunchSiteFilter.module.css";

const LaunchSiteFilter = (props) => {
  const { setLaunchSiteFilter } = props;

  return (
    <form
      onChange={(event) => setLaunchSiteFilter(event.target.value)}
      className={styles.launchSite__filter}
    >
      <label htmlFor="filter_by_launchSite">Filter By Launch Site: </label>
      <input type="radio" name="site_selection" value="Kwajalein Atoll" />
      <label>Kwajalein Atoll</label>
      <input type="radio" name="site_selection" value="Cape Canaveral" />
      <label>Cape Canaveral</label>
      <input type="radio" name="site_selection" value="Kennedy Space Center" />
      <label>Kennedy Space Center</label>
      <input type="radio" name="site_selection" value="" defaultChecked />
      <label>Show All</label>
    </form>
  );
};

export default LaunchSiteFilter;
