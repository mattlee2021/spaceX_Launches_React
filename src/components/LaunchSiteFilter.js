import { useState } from "react";
/**
 * This component displays a form with radio buttons so that a user can filter by one of three SpaceX
 * launch sites. I chose to use radio buttons because the SpaceX GraphQL find query takes a single 
 * String input. Thus, I only one input should be send to the API and radio buttons let a user choose 
 * mutually exclusive options. in the return block, I am passing the value of each input tag to App.js,
 * which then sends the value to the GraphQL API. The SpaceX GraphQL API will not filter by a specific 
 * input when an empty string is passed in, so for the Show All option, the value is an empty String.
 */


export function LaunchSiteFilter(props) {
    const [launchSite_state, launchSite_set] = useState("");
    const { onSubmit } = props; 

    const handleChange_launchSite = (event) => {
        launchSite_set(event.target.value)

    }
    const onSubmit_launchSite = (event) => {
        event.preventDefault()
        onSubmit(launchSite_state);
    }

    return (
        <form onSubmit={onSubmit_launchSite} onChange={handleChange_launchSite} className="Filtering">
            <label htmlFor="filter_by_launchSite">Filter By Launch Site: </label>
            <input type="radio" name="site_selection" value="Kwajalein Atoll"></input>
            <label>Kwajalein Atoll</label>
            <input type="radio" name="site_selection" value="Cape Canaveral"></input>
            <label>Cape Canaveral</label>
            <input type="radio" name="site_selection" value="Kennedy Space Center"></input>
            <label>Kennedy Space Center</label>
            <input type="radio" name="site_selection" value=""></input> 
            <label>Show All</label>
            <button type="submit">Filter</button>
        </form>
    );
}

