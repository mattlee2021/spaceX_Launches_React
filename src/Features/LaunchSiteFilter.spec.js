import TestRenderer from "react-test-renderer";
import LaunchSiteFilter from "./LaunchSiteFilter";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("LaunchSiteFilter", () => {
  it("Should match the snapshot", () => {
    const testRenderer = TestRenderer.create(<LaunchSiteFilter />);
    testRenderer.toJSON();
    expect(testRenderer).toMatchSnapshot();
  });

  it("Should call setLaunchSiteFilter with Kwajalein Atoll", () => {
    const mockSetLaunchSiteFilter = jest.fn();
    render(<LaunchSiteFilter setLaunchSiteFilter={mockSetLaunchSiteFilter} />);
    const radioButton = screen.getByLabelText("Kwajalein Atoll");
    userEvent.click(radioButton);
    expect(mockSetLaunchSiteFilter).toHaveBeenCalledWith("Kwajalein Atoll");
  });

  it("Should call setLaunchSiteFilter with Cape Canaveral", () => {
    const mockSetLaunchSiteFilter = jest.fn();
    render(<LaunchSiteFilter setLaunchSiteFilter={mockSetLaunchSiteFilter} />);
    const radioButton = screen.getByLabelText("Cape Canaveral");
    userEvent.click(radioButton);
    expect(mockSetLaunchSiteFilter).toHaveBeenCalledWith("Cape Canaveral");
  });

  it("Should call setLaunchSiteFilter with Kennedy Space Center", () => {
    const mockSetLaunchSiteFilter = jest.fn();
    render(<LaunchSiteFilter setLaunchSiteFilter={mockSetLaunchSiteFilter} />);
    const radioButton = screen.getByLabelText("Kennedy Space Center");
    userEvent.click(radioButton);
    expect(mockSetLaunchSiteFilter).toHaveBeenCalledWith(
      "Kennedy Space Center"
    );
  });

  it("Should call setLaunchSiteFilter with Show All, when Show All is not currently selected", () => {
    const mockSetLaunchSiteFilter = jest.fn();
    render(<LaunchSiteFilter setLaunchSiteFilter={mockSetLaunchSiteFilter} />);
    userEvent.click(screen.getByLabelText("Kennedy Space Center"));
    const radioButton = screen.getByLabelText("Show All");
    userEvent.click(radioButton);
    expect(mockSetLaunchSiteFilter).toHaveBeenCalledWith("");
  });
});
