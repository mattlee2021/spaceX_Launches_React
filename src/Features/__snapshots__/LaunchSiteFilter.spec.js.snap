// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchSiteFilter Should match the snapshot 1`] = `
<form
  className="launchSite__filter"
  onChange={[Function]}
>
  <label
    htmlFor="filter_by_launchSite"
  >
    Filter By Launch Site: 
  </label>
  <input
    id="Kwajalein Atoll"
    name="site_selection"
    type="radio"
    value="Kwajalein Atoll"
  />
  <label
    htmlFor="Kwajalein Atoll"
  >
    Kwajalein Atoll
  </label>
  <input
    id="Cape Canaveral"
    name="site_selection"
    type="radio"
    value="Cape Canaveral"
  />
  <label
    htmlFor="Cape Canaveral"
  >
    Cape Canaveral
  </label>
  <input
    id="Kennedy Space Center"
    name="site_selection"
    type="radio"
    value="Kennedy Space Center"
  />
  <label
    htmlFor="Kennedy Space Center"
  >
    Kennedy Space Center
  </label>
  <input
    defaultChecked={true}
    id="Show All"
    name="site_selection"
    type="radio"
    value=""
  />
  <label
    htmlFor="Show All"
  >
    Show All
  </label>
</form>
`;
